lib/scripts.js

```diff
@@ -305,93 +316,51 @@ var scripts = {
    *
    * @param {Queue} queue The queue for the job
    * @param {Job} job The job
-   * @param {Boolean=false} renew Whether to renew to lock, meaning it will assume the job 
+   * @param {Boolean=false} renew Whether to renew to lock, meaning it will assume the job
    *    is already locked and just reset the lock expiration.
    * @param {Boolean=false} ensureActive Ensures that the job is in the 'active' state.
    */
-  takeLock: function(queue, job, token, renew, ensureActive){
-    // Ensures that the lock doesn't exist, or if it does, that we own it.
-    var ensureOwnershipCall = [
-      'local prevLock = redis.call("GET", KEYS[1])',
-      'if (prevLock and prevLock ~= ARGV[1]) then',
-      '  return 0',
-      'end'
-    ].join('\n');
-
-    // Ensures that the lock in the 'active' state.
-    var ensureActiveCall = [
-      // Note that while this is inefficient to run a O(n) traversal of the 'active' queue,
-      // it's highly likely that the job is within the first few elements of the active
-      // list at the time this call is used.
-      'local activeJobs = redis.call("LRANGE", KEYS[3], 0, -1)',
-      'local found = false',
-      'for _, job in ipairs(activeJobs) do',
-      '  if(job == ARGV[3]) then',
-      '     found = true',
-      '     break',
-      '  end',
-      'end',
-      'if (found == false) then',
-      '  return 0',
-      'end'
-    ].join('\n');
+  takeLock: function(queue, job, renew, ensureActive){
+    var lock = job.lock;
+    if (renew && !lock) {
+      throw new Error('Unable to renew nonexisting lock');
+    } else if (renew) {
+      return lock.extend(queue.LOCK_RENEW_TIME);
+    } else if (lock) {
+      return Promise.resolve(null);
+    }
 
-    var lockCall;
-    if (renew){
-      lockCall = 'redis.call("SET", KEYS[1], ARGV[1], "PX", ARGV[2])';
+    var redlock;
+    if (ensureActive) {
+      var keyVar = ['"', job.queue.toKey('active'), '"'].join('');
+      var argVar = ['"', job.jobId, '"'].join('');
+      var isJobInList = this._isJobInList(keyVar, argVar, 'if');
+      var lockAcquired = ['and redis.call("HSET", "', queue.toKey(job.jobId), '", "lockAcquired", "1")'].join('');
+      var success = 'then return 1 else return 0 end';
+      var opts = {
+        lockScript: function(lockScript) {
+          return [
+            isJobInList,
+            lockScript.replace('return', 'and'),
+            lockAcquired,
+            success
+          ].join('\n');
+        }
+      };
+      redlock = new Redlock(queue.clients, _.extend(opts, queue.redlock));
     } else {
-      lockCall = 'redis.call("SET", KEYS[1], ARGV[1], "PX", ARGV[2], "NX")';
+      redlock = new Redlock(queue.clients, queue.redlock);
```

Comment:
Yeah, that's what I mean - overwrite the private variable. Lock the version down in package.json (eg `"redlock": "2.0.2"`) to make sure that variable won't change on us. Then in code:

```
var redlock = new Redlock(..)
redlock.lockScript = redlock.lockScript.replace(...)
```

This is much safer because it still uses the real `redlock` module, and a locked version that can't update. It's clearer in package.json so other devs can see what version we're using. You can even add a comment in package.json (http://stackoverflow.com/questions/14221579/how-do-i-add-comments-to-package-json-for-npm-install) to indicate why it's locked to a specific version.

Pointing this to `nextorigin/node-redlock` is dangerous because if you or anyone on your team accidentally pushed to that master branch, it'll break Bull for **all** users. I trust that you wouldn't do that, but it's unnecessary risk.

cc @manast